compiler/runtime/OMRRelocationRuntime.hpp

```diff
@@ -0,0 +1,102 @@
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef RELOCATION_RUNTIME_INCL
+#define RELOCATION_RUNTIME_INCL
+
+#include "omrcfg.h"
+#include "omr.h"
+#include <map>
+#include <assert.h>
+#include "codegen/Relocation.hpp"
+#include "runtime/Runtime.hpp"
+#include "env/OMRCPU.hpp" 
+
+#ifndef OMR_RELOCATION_RUNTIME_CONNECTOR
+#define OMR_RELOCATION_RUNTIME_CONNECTOR
+namespace OMR { class RelocationRuntime; }
+namespace OMR { typedef OMR::RelocationRuntime RelocationRuntimeConnector; }
+#endif
+
+#ifndef OMR_SHARED_RELOCATION_RUNTIME_CONNECTOR
+#define OMR_SHARED_RELOCATION_RUNTIME_CONNECTOR
+namespace OMR { class SharedCacheRelocationRuntime; }
+namespace OMR { typedef OMR::SharedCacheRelocationRuntime SharedCacheRelocationRuntimeConnector; }
+#endif
+
+
+namespace TR { 
+   class RelocationRecord;
+   class RelocationTarget;
+   class RelocationRecordBinaryTemplate;
+   class AOTMethodHeader;
+   class RelocationRuntime;
+   class JitConfig;
+   }
+
+namespace OMR{
+
+class OMR_EXTENSIBLE RelocationRuntime {
+   public:
+      TR_ALLOC(TR_Memory::Relocation)
+      void * operator new(size_t, TR::JitConfig *);
+      RelocationRuntime(TR::JitConfig *jitCfg);
+      TR::RelocationRuntime* self();
+      TR::RelocationTarget *reloTarget()                           { return _reloTarget; }
+      TR_Memory *trMemory()                                       { return _trMemory; }
+      TR::AOTMethodHeader *aotMethodHeaderEntry()                  { return _aotMethodHeaderEntry; }
+      TR::Compilation* comp()                                    { return _comp; }
+      // current main entry point
+      void *prepareRelocateAOTCodeAndData(TR::AOTMethodHeader *cacheEntry, void* code);
+   private:
+      void relocateAOTCodeAndData(U_8 *codeStart);
+
+      typedef enum {
+         RelocationNoError = 1,
+      } TR_AotRelocationCleanUp;
+      TR_AotRelocationCleanUp _relocationStatus;
+      void relocationFailureCleanup();
+   protected:
+      TR_Memory *_trMemory;
+      TR::RelocationTarget *_reloTarget;
+      TR::AOTMethodHeader *_aotMethodHeaderEntry;
+      TR::Compilation *_comp;
+};
+}
+namespace OMR
+{
+class OMR_EXTENSIBLE SharedCacheRelocationRuntime : public OMR::RelocationRuntime {
```

Comment:
You need to be careful here because you're essentially merging the component classes of two extensible hierarchies.  Do you mean:
```
class OMR_EXTENSIBLE SharedCacheRelocationRuntime : public TR::RelocationRuntime {
```
If so, then RelocationRuntime should live in a separate file.

Reply:
Hi, thank you for the comment, I've updated code accordingly